# Analyzing the startup that sells food products. Investigation of user behavior for the company's app.

In this project we have to prepare a report with the users behaviour for the company's app. We need to find out the share of users that go from main screen to the purhase and 

**Description of the data**

children - the number of children in the family
   
days_employed - how long the customer has been working
   
dob_years -  the customer’s age
   
education - the customer’s education level

education_id - identifier for the customer’s education

family_status - the customer’s marital status

family_status_id - identifier for the customer’s marital status

gender - the customer’s gender

income_type - the customer’s income type

debt - whether the customer has ever defaulted on a loan

total_income - monthly income

purpose -  reason for taking out a loan


## Open the data file and have a look at the general information. 

In [1]:
import pandas as pd 
credit_scoring = pd.read_csv('/datasets/credit_scoring_eng.csv')


In [2]:
credit_scoring.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [3]:
credit_scoring.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

In [4]:
credit_scoring.describe(include = ['object'])

education family_status gender income_type           purpose
count                 21525         21525  21525       21525             21525
unique                   15             5      3           8                38
top     secondary education       married      F    employee  wedding ceremony
freq                  13750         12380  14236       11119               797

In [67]:
credit_scoring['children'].describe()

count    21454.000000
mean         0.480563
std          0.756069
min          0.000000
25%          0.000000
50%          0.000000
75%          1.000000
max          5.000000
Name: children, dtype: float64

In [68]:
credit_scoring['children'].value_counts()


0    14091
1     4855
2     2128
3      330
4       41
5        9
Name: children, dtype: int64

In [69]:
credit_scoring['dob_years'].value_counts().sort_index()

0     101
19     14
20     51
21    111
22    183
23    252
24    264
25    357
26    408
27    493
28    503
29    544
30    537
31    559
32    509
33    581
34    601
35    616
36    554
37    536
38    597
39    572
40    607
41    605
42    596
43    512
44    545
45    496
46    472
47    477
48    536
49    508
50    513
51    446
52    484
53    459
54    476
55    443
56    483
57    456
58    454
59    443
60    374
61    354
62    348
63    269
64    260
65    193
66    182
67    167
68     99
69     85
70     65
71     56
72     33
73      8
74      6
75      1
Name: dob_years, dtype: int64

In [70]:
credit_scoring.head()

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0         40620   
1            married                 0      F    employee     0         17932   
2            married                 0      M    employee     0         23341   
3            married                 0      M    employee     0         42820   
4  civil partnership                 1      F     retiree     0         25378   

                   purpose income_category age_category    purpose_cat  
0    purchase of the house            high       adults       property  
1             car purchase    lower_middle       adults            car  
2    purchase of the house    upper_middle       adults       property  
3  supplementary education            high       adults      education  
4        to have a wedding    upper_middle       adults  special_event

In [71]:
credit_scoring.tail()

children  days_employed  dob_years            education  education_id  \
21520         1    4529.316663         43  secondary education             1   
21521         0  343937.404131         67  secondary education             1   
21522         1    2113.346888         38  secondary education             1   
21523         3    3112.481705         38  secondary education             1   
21524         2    1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income               purpose income_category age_category  \
21520         35966  housing transactions            high       adults   
21521         24959     purchase of a car    upper_middle       senior   
21522         14347              property             low       adults   
21523         39054     buying my own car            high       adults   
21524         13127          to buy a car             low       adults   

      purpose_cat  
21520    property  
21521         car  
21522    property  
21523         car  
21524         car

From the first look at our data file, we see that there are 21525 rows and we have 12 columns. We also see that some columns have missing data (income column), weird data (children (min = -1, max = 20),age (0), days_employed, missing values for total_income). We can't say for sure, if the data in income_column missing on purpose or it is a technical mistake. As we see, there is a weird information about children.I suppose that we have a typo in children column, so i decided to replace (-1) by (1) and (20) by (2). Also we see that we have the same amount of missing values in columns days_employed and total_income. For our research we obviously need to fill missing values in the column (total_income), but we don't need as the must - column (days_employed), so in this research I decided not to fix it, but to pass information to the data holders, that something went wrong with days_employed data.

## Data preprocessing

### Processing missing values

In [10]:
credit_scoring.isnull().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [11]:
for i in credit_scoring:
    if credit_scoring[i].isnull().sum() > 0:
        print(i)

days_employed
total_income


In [12]:
credit_scoring['children'].replace([-1], 1, inplace=True)

In [13]:
credit_scoring['children'].replace([20], 2, inplace=True)

I ve decided to replace values because in my opinion a typo was made, it is impossible to have -1 or in common ways it is very rare to have 20.

In [14]:
credit_scoring['children'].value_counts()

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

In [15]:
credit_scoring['education'].unique()

array(["bachelor's degree", 'secondary education', 'Secondary Education',
       'SECONDARY EDUCATION', "BACHELOR'S DEGREE", 'some college',
       'primary education', "Bachelor's Degree", 'SOME COLLEGE',
       'Some College', 'PRIMARY EDUCATION', 'Primary Education',
       'Graduate Degree', 'GRADUATE DEGREE', 'graduate degree'],
      dtype=object)

In [16]:
credit_scoring['education'] = credit_scoring['education'].str.lower()

In [17]:
credit_scoring['education'].value_counts()

secondary education    15233
bachelor's degree       5260
some college             744
primary education        282
graduate degree            6
Name: education, dtype: int64

In [18]:
median_income = credit_scoring.groupby('income_type')['total_income'].median()

In [19]:
median_income

income_type
business                       27577.2720
civil servant                  24071.6695
employee                       22815.1035
entrepreneur                   79866.1030
paternity / maternity leave     8612.6610
retiree                        18962.3180
student                        15712.2600
unemployed                     21014.3605
Name: total_income, dtype: float64

In [20]:
credit_scoring['total_income'] = credit_scoring['total_income'].fillna(credit_scoring.groupby('income_type')['total_income'].transform('median'))

In [21]:
credit_scoring['total_income'].isnull().value_counts()

False    21525
Name: total_income, dtype: int64

I found out that there is missing data in the column Total income, so i filled it with the median values grouped by income type. Then I checked if the code worked well. Now we can procced for the further investigation

### Conclusion

First of all, I fixed the strange information about children by replacing -1  to 1 and 20 children to 2. Also we see that there is the same text in the 'Education', but written in the different ways. So I used 'lower' method to rewrite this columns so that we have the same types of education to analyse. Then I found the median value for total income in several groups sorted by income_type (employee, enterpreneur, retiree,unemployed and else), after that I replaced the missing values with that median income for all the specified groups.

### Data type replacement

In [22]:
credit_scoring['days_employed'].describe()

count     19351.000000
mean      63046.497661
std      140827.311974
min      -18388.949901
25%       -2747.423625
50%       -1203.369529
75%        -291.095954
max      401755.400475
Name: days_employed, dtype: float64

In [23]:
credit_scoring['days_employed'] = credit_scoring['days_employed'].abs()

In [24]:
credit_scoring['days_employed'].describe()

count     19351.000000
mean      66914.728907
std      139030.880527
min          24.141633
25%         927.009265
50%        2194.220567
75%        5537.882441
max      401755.400475
Name: days_employed, dtype: float64

From the first look at our column "Days employed" we noticed that there is a really problematic data. We saw negative values, we saw there a really huge max value ( if we consider that data is in years format = than the max value is 1100 years of working). Using abs() we can improve data fixing negative values. But I don't still see the possibility to use this column. I think we should leave a comment that we must clean the data with someone who filled in the column. But for this research we actually don't need it.

In [25]:
credit_scoring['days_employed'] = credit_scoring['days_employed'].fillna(credit_scoring.groupby(['education_id', 'income_type'])['days_employed'].transform('median'))

In [26]:
credit_scoring['days_employed'].isnull().value_counts()

False    21525
Name: days_employed, dtype: int64

In [27]:
credit_scoring.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [28]:
credit_scoring['gender'].describe()

count     21525
unique        3
top           F
freq      14236
Name: gender, dtype: object

In [29]:
credit_scoring['gender'].unique()

array(['F', 'M', 'XNA'], dtype=object)

We see that in this gender we have a XNA value. So i want to replace it with Female or Male. To do so I'll try to understand the most popular value in several groups

In [30]:
marital_gender=pd.pivot_table(credit_scoring,index='gender',columns='family_status',values='family_status_id',aggfunc='count',margins=True).reset_index()
marital_gender

family_status gender  civil partnership  divorced  married  unmarried  \
0                  F             2868.0     936.0   7795.0     1732.0   
1                  M             1308.0     259.0   4585.0     1081.0   
2                XNA                1.0       NaN      NaN        NaN   
3                All             4177.0    1195.0  12380.0     2813.0   

family_status  widow / widower    All  
0                        905.0  14236  
1                         55.0   7288  
2                          NaN      1  
3                        960.0  21525

In [31]:
education_gender=pd.pivot_table(credit_scoring,index='gender',columns='education',values='education_id',aggfunc='count',margins=True).reset_index()
education_gender

education gender  bachelor's degree  graduate degree  primary education  \
0              F             3547.0              2.0              163.0   
1              M             1713.0              4.0              119.0   
2            XNA                NaN              NaN                NaN   
3            All             5260.0              6.0              282.0   

education  secondary education  some college    All  
0                      10064.0         460.0  14236  
1                       5169.0         283.0   7288  
2                          NaN           1.0      1  
3                      15233.0         744.0  21525

Looking at the overall procent of Female customers over Male customers and the several tables grouped by marital status and education, I've decided to replace XNA with Female, as it shouldn't change overall statistic

In [32]:
credit_scoring['gender'] = credit_scoring['gender'].replace(['XNA'],'F')

In [33]:
credit_scoring['gender'].unique()

array(['F', 'M'], dtype=object)

In [34]:
def make_int(i):
    try:
        return int(i)
    except:
        return pd.np.nan

credit_scoring['total_income'] = credit_scoring['total_income'].map(make_int)

In [35]:
credit_scoring.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  int64  
 11  purpose           21525 non-null  object 
dtypes: float64(1), int64(6), object(5)
memory usage: 2.0+ MB


### Conclusion

In this part we figured out that there was not a correct value in the gender column,so I've replaced it with Female, because the procentage of Female is higher than Male over all data and by different groups too (i took marital status, education for checking)
Also I changed data type for total_income, to make this column more convenient to look through.

### Processing duplicates

In [36]:
credit_scoring.duplicated().value_counts()

False    21454
True        71
dtype: int64

In [37]:
credit_scoring.duplicated().sum() / len(credit_scoring) *100

0.32984901277584205

In [38]:
credit_scoring.drop_duplicates(inplace=True)

In [39]:
credit_scoring.duplicated().value_counts()

False    21454
dtype: int64

The precentage of duplicates turned out to be small, so we removed it to improve data. 

### Conclusion

We see that there were duplicated rows, we counted the percentage of them and it seemed to be a good idea to remove these duplicated rows.

### Categorizing Data

In [40]:
credit_scoring['total_income'].describe()

count     21454.000000
mean      26450.742799
std       15709.964409
min        3306.000000
25%       17219.500000
50%       22815.000000
75%       31331.000000
max      362496.000000
Name: total_income, dtype: float64

In [41]:
def income_categorize(row):
    if row[10] <= 17247.708:
        return 'low'
    elif row[10] > 17247.708 and row[10] <= 22815.1035:
        return 'lower_middle'
    elif row[10] > 22815.1035 and row[10] <= 31287.991:
        return 'upper_middle'
    else:
        return 'high'

In [42]:
credit_scoring['income_category'] = credit_scoring.apply(income_categorize, axis=1)

In [43]:
credit_scoring.head()

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0         40620   
1            married                 0      F    employee     0         17932   
2            married                 0      M    employee     0         23341   
3            married                 0      M    employee     0         42820   
4  civil partnership                 1      F     retiree     0         25378   

                   purpose income_category  
0    purchase of the house            high  
1             car purchase    lower_middle  
2    purchase of the house    upper_middle  
3  supplementary education            high  
4        to have a wedding    upper_middle

In [44]:
credit_scoring['income_category'].isnull().value_counts()

False    21454
Name: income_category, dtype: int64

In [45]:
credit_scoring['dob_years'].describe()

count    21454.000000
mean        43.271231
std         12.570822
min          0.000000
25%         33.000000
50%         42.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

In [46]:
def age_categories(row):
    if row[2] <= 14:
        return 'children'
    elif row[2] > 14 and row[2] <= 24:
        return 'youth'
    elif row[2] > 24 and row[2] <= 64:
        return 'adults'
    elif row[2] >64:
        return 'senior'
    else:
        return 'other'

In [47]:
credit_scoring['age_category'] = credit_scoring.apply(age_categories, axis=1)

In [48]:
credit_scoring[credit_scoring['age_category']== 'other']

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose, income_category, age_category]
Index: []

In [49]:
credit_scoring.age_category.unique()

array(['adults', 'senior', 'youth', 'children'], dtype=object)

In [50]:
credit_scoring.purpose.unique()

array(['purchase of the house', 'car purchase', 'supplementary education',
       'to have a wedding', 'housing transactions', 'education',
       'having a wedding', 'purchase of the house for my family',
       'buy real estate', 'buy commercial real estate',
       'buy residential real estate', 'construction of own property',
       'property', 'building a property', 'buying a second-hand car',
       'buying my own car', 'transactions with commercial real estate',
       'building a real estate', 'housing',
       'transactions with my real estate', 'cars', 'to become educated',
       'second-hand car purchase', 'getting an education', 'car',
       'wedding ceremony', 'to get a supplementary education',
       'purchase of my own house', 'real estate transactions',
       'getting higher education', 'to own a car', 'purchase of a car',
       'profile education', 'university education',
       'buying property for renting out', 'to buy a car',
       'housing renovation', 'going

In [51]:
import nltk
from nltk.stem import WordNetLemmatizer
wordnet_lemma = WordNetLemmatizer()

In [52]:
def lemmatize_text(text):
    return [wordnet_lemma.lemmatize(w) for w in nltk.word_tokenize(text)]

In [53]:
purpose_lemmatized = credit_scoring.purpose.apply(lemmatize_text)

In [54]:
purpose_lemmatized

0        [purchase, of, the, house]
1                   [car, purchase]
2        [purchase, of, the, house]
3        [supplementary, education]
4            [to, have, a, wedding]
                    ...            
21520        [housing, transaction]
21521        [purchase, of, a, car]
21522                    [property]
21523        [buying, my, own, car]
21524             [to, buy, a, car]
Name: purpose, Length: 21454, dtype: object

In [55]:
property_category=['house', 'estate', 'property', 'housing']
education_category=['education', 'university', 'educated'] 
special_event_category=['wedding', 'ceremony']
car_category=['car']

In [56]:
any(word in purpose_lemmatized for word in property_category)

False

In [57]:
def lemmatization_func(line):
    lemmatized=lemmatize_text(line)
    if any(word in lemmatized for word in property_category):
        return 'property'
    elif any(word in lemmatized for word in education_category):
        return 'education'
    elif any(word in lemmatized for word in special_event_category):
        return 'special_event'
    elif any(word in lemmatized for word in car_category):
        return 'car'
    else:
        return 'other'

In [58]:
credit_scoring['purpose_cat']=credit_scoring['purpose'].apply(lemmatization_func)

In [59]:
credit_scoring['purpose_cat'].value_counts()

property         10811
car               4306
education         4013
special_event     2324
Name: purpose_cat, dtype: int64

In [60]:
credit_scoring[credit_scoring['purpose_cat']== 'other']

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose, income_category, age_category, purpose_cat]
Index: []

### Conclusion

Analysing data about credit purposes, we saw that there was a reall mess there, so using the lemmatization function i devided all the information into 4 groups ( property, car, education and special event)

## Answer these questions

- Is there a relation between having kids and repaying a loan on time?

In [61]:
credit_scoring.groupby(['children'])['debt'].mean().reset_index()

children      debt
0         0  0.075438
1         1  0.091658
2         2  0.094925
3         3  0.081818
4         4  0.097561
5         5  0.000000

### Conclusion

Drawing conclusions of connection between having kids and repaying loan on time we can see the followinf dependence:
People with 1,2 or 4 children had the most problematic experience to return the loan on time, in the other hand, we see that people with 0 kids were the most accurate with returning money. I don't think that we should pay attention to group with 5 children, because they are a really little procent of the whole list of data.

- Is there a relation between marital status and repaying a loan on time?

In [62]:
marital_status=pd.pivot_table(credit_scoring,index='family_status',columns='debt',values='family_status_id',aggfunc='count',margins=True).reset_index()
marital_status

debt      family_status      0     1    All
0     civil partnership   3763   388   4151
1              divorced   1110    85   1195
2               married  11408   931  12339
3             unmarried   2536   274   2810
4       widow / widower    896    63    959
5                   All  19713  1741  21454

In [63]:
marital_status['default_rate']=(marital_status[1]/marital_status['All'])*100
marital_status

debt      family_status      0     1    All  default_rate
0     civil partnership   3763   388   4151      9.347145
1              divorced   1110    85   1195      7.112971
2               married  11408   931  12339      7.545182
3             unmarried   2536   274   2810      9.750890
4       widow / widower    896    63    959      6.569343
5                   All  19713  1741  21454      8.115037

### Conclusion

Analysing the data given, we can refer the group with widow family_status to the one who return loans mostly in time, and the unmarried group has the biggest procent of loans that were not returned

- Is there a relation between income level and repaying a loan on time?

In [64]:
income_level=pd.pivot_table(credit_scoring,index='income_category',columns='debt',values='total_income',aggfunc='count',margins=True).reset_index()
income_level

debt income_category      0     1    All
0               high   4995   386   5381
1                low   4955   427   5382
2       lower_middle   4979   483   5462
3       upper_middle   4784   445   5229
4                All  19713  1741  21454

In [65]:
income_level['default_rate']=(income_level[1]/income_level['All'])*100
income_level

debt income_category      0     1    All  default_rate
0               high   4995   386   5381      7.173388
1                low   4955   427   5382      7.933854
2       lower_middle   4979   483   5462      8.842915
3       upper_middle   4784   445   5229      8.510231
4                All  19713  1741  21454      8.115037

### Conclusion

Following the information above we can describe the situation with credit loans so:
the income level slightly affects the on-time repayment. We can say that the less reliable category is clients with lower-middle income, the most reliable are those with high income.

- How do different loan purposes affect on-time repayment of the loan?

In [66]:
credit_scoring.groupby(['purpose_cat'])['debt'].mean().reset_index()

purpose_cat      debt
0            car  0.093590
1      education  0.092200
2       property  0.072334
3  special_event  0.080034

### Conclusion

Considering the information that we see in the table most of the problems with returning money is with the credits to buy a car and the education credit, the most reliable is property category.

## General conclusion

Summing up all the information we can see following. We had a not really big data for research, there was only  apbout 21000 lines. We had information about children quantity, marrital status, purpose of loan, type of income and total income. Some of the information was missing ( for example total_income), some had typo (information about children). After fixing that problems, we categorized clients by age, by income, by purposes of credit. The most common group of people, sliced by age are adults, the most popular aim to take a credit - loans for property, the widest category by family status are married people.
Now we face that the reliable category of clients are widows or married people, people with high income and people who take a loan to buy property. We should think over more when deciding to approve a loan to people with lower-middle total income unmarried or people in civil partnership status. 